### PR TITLE
Renderers

### PR DESCRIPTION
**Pull request by eugenebokhan**

### CODE REVIEW

File: Alloy/Encoders/BoundingBoxesRenderer.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
+//
+//  BoundingBoxesRenderer.swift
```
Comment (s1ddok) [inline Alloy/Encoders/BoundingBoxesRenderer.swift]: I don't think `/Encoders` is the right location for this file

File: Alloy/Encoders/BoundingBoxesRenderer.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+//  Alloy
+//
+//  Created by Eugene Bokhan on 22/04/2019.
+//
+
+import Metal
+import simd
+
+@available(iOS 11.3, tvOS 11.3, macOS 10.13, *)
+final public class BoundingBoxesRenderer {
```
Comment (s1ddok) [inline Alloy/Encoders/BoundingBoxesRenderer.swift]: I really think this is over-complexity. Why not declare a cumulative index buffer for rectangle and render it in a single draw call?

File: Alloy/Renderers/RectangleRenderer.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ///
+    /// - Parameters:
+    ///   - renderEncoder: container to put the rendering work into.
+    public func draw(using renderEncoder: MTLRenderCommandEncoder) {
+        // Push a debug group allowing us to identify render commands in the GPU Frame Capture tool.
+        renderEncoder.pushDebugGroup("Draw Rectangle Geometry")
+
+        // Set render command encoder state.
+        renderEncoder.setRenderPipelineState(self.renderPipelineState)
+        // Set any buffers fed into our render pipeline.
```
Comment (s1ddok) [inline Alloy/Renderers/RectangleRenderer.swift]: style

File: Alloy/Renderers/RectangleRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                     commandBuffer: MTLCommandBuffer) throws {
+        // Check render target.
+        guard let renderTarget = renderPassDescriptor.colorAttachments[0].texture
+        else { throw Errors.missingRenderTarget }
+        guard renderTarget.usage.contains(.renderTarget)
+        else { throw Errors.wrongRenderTargetTextureUsage }
+
+        // Draw.
+        commandBuffer.render(descriptor: renderPassDescriptor) { renderEncoder in
+            self.draw(using: renderEncoder)
```
Comment (s1ddok) [inline Alloy/Renderers/RectangleRenderer.swift]: there is no need in second closure

File: Alloy/Renderers/DebugRenderer.swift [hunk 5]
```diff
@@ -0,0 +1,21 @@
+//
+//  DebugRenderer.swift
```
Comment (s1ddok) [inline Alloy/Renderers/DebugRenderer.swift]: it doesnt belong here

File: Alloy/Renderers/BoundingBoxesRenderer.swift [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+    private func calculateComponentLine(bboxRect: CGRect,
+                                        lineWidth: CGFloat,
+                                        textureWidth: CGFloat,
+                                        textureHeight: CGFloat,
+                                        for componentType: ComponentType) -> Line {
+        let horizontalWidth = lineWidth / textureHeight
+        let verticalWidth = lineWidth / textureWidth
+
+        let rect = CGRect(x: -1 + bboxRect.minX * 2,
```
Comment (s1ddok) [inline Alloy/Renderers/BoundingBoxesRenderer.swift]: style

File: Alloy/Shaders/ShaderStructures.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+typedef struct Rectangle {
+    vector_float2 topLeft;
+    vector_float2 bottomLeft;
+    vector_float2 topRight;
+    vector_float2 bottomRight;
+    vector_float4 fillColor;
+} Rectangle;
+
+typedef struct Line {
```
Comment (s1ddok) [inline Alloy/Shaders/ShaderStructures.h]: unsafe memory layout

File: Alloy/Renderers/RectangleRenderer.swift [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    /// Draw a rectangle in a target texture.
+    ///
+    /// - Parameters:
+    ///   - renderPassDescriptor: render pass descriptor to be used.
+    ///   - commandBuffer: command buffer to put the rendering work items into.
+    public func draw(renderPassDescriptor: MTLRenderPassDescriptor,
+                     commandBuffer: MTLCommandBuffer) throws {
+        // Check render target.
+        guard let renderTarget = renderPassDescriptor.colorAttachments[0].texture
```
Comment (s1ddok) [inline Alloy/Renderers/RectangleRenderer.swift]: #if DEBUG

File: Alloy/Shaders/Shaders.metal [hunk 9]
```diff
@@ -179,3 +180,69 @@ kernel void mean(texture2d<half, access::sample> input_texture [[ texture(0) ]],
...
+}
+
+// MARK: - Lines Rendering
+
+inline float2 perp(float2 in) {
+    return float2(-in.y, in.x);
+}
+
+vertex VertexOut linesVertex(constant Line *lines [[ buffer(0) ]],
+                       uint vertexId [[vertex_id]],
```
Comment (s1ddok) [inline Alloy/Shaders/Shaders.metal]: style

File: Alloy/Renderers/RectangleRenderer.swift [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+final public class RectangleRenderer {
+
+    // MARK: - Properties
+
+    /// Rectangle fill color.
+    public var color: CGColor = .black
+    /// Rectrangle in a normalized coodrinate system to draw.
+    public var normalizedRect: CGRect = .zero
+
+    private var renderPipelineState: MTLRenderPipelineState!
```
Comment (s1ddok) [inline Alloy/Renderers/RectangleRenderer.swift]: why optional?

File: Alloy/Renderers/RectangleRenderer.swift [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import Metal
+import simd
+
+@available(iOS 11.3, tvOS 11.3, macOS 10.13, *)
+final public class RectangleRenderer {
+
+    // MARK: - Properties
+
+    /// Rectangle fill color.
+    public var color: CGColor = .black
```
Comment (s1ddok) [inline Alloy/Renderers/RectangleRenderer.swift]: it can be easier

File: Alloy/Renderers/BoundingBoxesRenderer.swift [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+//  Created by Eugene Bokhan on 22/04/2019.
+//
+
+import Metal
+import simd
+
+@available(iOS 11.3, tvOS 11.3, macOS 10.13, *)
+final public class BoundingBoxesRenderer {
+
+    private enum ComponentType {
```
Comment (s1ddok) [inline Alloy/Renderers/BoundingBoxesRenderer.swift]: questionable

File: Alloy/Renderers/RectangleRenderer.swift (hunk 4 above)
Comment (eugenebokhan) [reply Alloy/Renderers/RectangleRenderer.swift]: I didn't quiet get it. Do we need to remove `draw(using renderEncoder: MTLRenderCommandEncoder)` at all?

File: Alloy/Renderers/RectangleRenderer.swift (hunk 3 above)
Comment (eugenebokhan) [reply Alloy/Renderers/RectangleRenderer.swift]: what's wrong here?

File: Alloy/Renderers/BoundingBoxesRenderer.swift (hunk 6 above)
Comment (eugenebokhan) [reply Alloy/Renderers/BoundingBoxesRenderer.swift]: what's wrong here?

File: Alloy/Shaders/Shaders.metal [hunk 13]
```diff
@@ -179,3 +180,67 @@ kernel void mean(texture2d<half, access::sample> input_texture [[ texture(0) ]],
...
+
+    float2 vector = startPoint - endPoint;
+    float2 perpendicularVector = perpendicular(normalize(vector));
+    float halfWidth = line.width / 2;
+
+    const float2 vertexPositions[] = {
+        startPoint, endPoint,
+        startPoint, endPoint,
+    };
+    const float offsetFactors[] = {
```
Comment (s1ddok) [inline Alloy/Shaders/Shaders.metal]: why not create only one stack array?

File: Alloy/Renderers/RectangleRenderer.swift (hunk 3 above)
Comment (s1ddok) [reply Alloy/Renderers/RectangleRenderer.swift]: empty line after comments

File: Alloy/MTLBuffer+Extensions.swift [hunk 14]
```diff
@@ -17,7 +17,7 @@ public extension MTLBuffer {
         else { return nil }
         
         #if DEBUG
-        if self.length >= MemoryLayout<T>.stride {
+        if self.length < MemoryLayout<T>.stride {
```
Comment (s1ddok) [inline Alloy/MTLBuffer+Extensions.swift]: It is fixed in current master, I think it should be removed from PR

File: Alloy/Renderers/BoundingBoxesRenderer.swift [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    public enum Errors: Error {
+        case wrongRenderTargetTextureUsage
+        case missingRenderTarget
+    }
+
+    // MARK: - Properties
+
+    /// Rectrangles in a normalized coodrinate system describing bounding boxes.
+    public var normalizedRects: [CGRect] = []
+    /// Rrefered fill color of the bounding boxes. Red is default.
```
Comment (s1ddok) [inline Alloy/Renderers/BoundingBoxesRenderer.swift]: that's not the fill color

File: Alloy/Renderers/BoundingBoxesRenderer.swift [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    public var color: vector_float4 = .init(1, 0, 0, 1) {
+        didSet {
+            self.linesRenderer.color = self.color
+        }
+    }
+    /// Prefered line width of the bounding boxes in pixels. 20 is default.
+    public var lineWidth: Int = 20
+    /// Size of the render target texture.
+    public var renderTargetSize: MTLSize = .zero
```
Comment (s1ddok) [inline Alloy/Renderers/BoundingBoxesRenderer.swift]: why do we need to know it?

File: Alloy/Renderers/LinesRenderer.swift [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    /// Render lines in a target texture.
+    ///
+    /// - Parameters:
+    ///   - renderPassDescriptor: Render pass descriptor to be used.
+    ///   - commandBuffer: Command buffer to put the rendering work items into.
+    public func render(renderPassDescriptor: MTLRenderPassDescriptor,
+                       commandBuffer: MTLCommandBuffer) throws {
+        #if DEBUG
+        // Check render target.
```
Comment (s1ddok) [inline Alloy/Renderers/LinesRenderer.swift]: I don't think this check is needed, Metal does it internally with Metal API Validation

File: Alloy/Shaders/Shaders.metal [hunk 18]
```diff
@@ -179,3 +180,102 @@ kernel void mean(texture2d<half, access::sample> input_texture [[ texture(0) ]],
...
+struct VertexOut {
+    float4 position [[ position ]];
+    half4 fillColor;
+};
+
+float2 perpendicular(float2 vector) {
+    return float2(-vector.y, vector.x);
+}
+
+float2 convertToMetalCoordinateSystem(float2 vector) {
```
Comment (s1ddok) [inline Alloy/Shaders/Shaders.metal]: I would call it screen space

File: Alloy/Shaders/ShaderStructures.h [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+#include <simd/simd.h>
+
+typedef struct Rectangle {
+    vector_float2 topLeft;
+    vector_float2 bottomLeft;
+    vector_float2 topRight;
+    vector_float2 bottomRight;
+} Rectangle;
+
+typedef struct Line {
```
Comment (s1ddok) [inline Alloy/Shaders/ShaderStructures.h]: I don't think storing that much is nice idea. I would just introduce PolyLineRenderer that have uniform width across its segments and can draw a series of points

File: Alloy/Shaders/Shaders.metal [hunk 20]
```diff
@@ -179,3 +180,102 @@ kernel void mean(texture2d<half, access::sample> input_texture [[ texture(0) ]],
     }
 
 }
+
+// MARK: - Rendering
+
+struct VertexOut {
+    float4 position [[ position ]];
+    half4 fillColor;
```
Comment (s1ddok) [inline Alloy/Shaders/Shaders.metal]: why store fillColor if it is not used

File: Alloy/Renderers/BoundingBoxesRenderer.swift (hunk 15 above)
Comment (eugenebokhan) [reply Alloy/Renderers/BoundingBoxesRenderer.swift]: I'll change it to `... border color of ...`.

File: Alloy/Renderers/BoundingBoxesRenderer.swift (hunk 16 above)
Comment (eugenebokhan) [reply Alloy/Renderers/BoundingBoxesRenderer.swift]: `renderTargetSize` is used to calculate vertical & horizontal widths of line components of the bbox.

**Pull request merged**